crypto/params.c

```diff
@@ -0,0 +1,287 @@
+        const size_t eb = p->buffer_size - sz;
+
+        if (is_endian.little) {
+            neg = sign && (((unsigned char *)val)[p->buffer_size - 1]
+                           & 0x80);
+            memcpy(p->buffer, val, sz);
+            memset(((unsigned char *)p->buffer) + sz, neg ? 0xff : 0, eb);
+        } else {
+            neg = sign && (*(unsigned char *)val & 0x80);
+            memset(p->buffer, neg ? 0xff : 0, eb);
+            memcpy(((unsigned char *)p->buffer) + eb, val, sz);
+        }
+    } else {                            /* Narrowing */
+        if (is_endian.little)
+            memcpy(p->buffer, val, p->buffer_size);
+        else
+            memcpy(p->buffer, ((unsigned char *)val) + (sz - p->buffer_size),
+                   p->buffer_size);
+    }
+    if (p->return_size != NULL)
+        *p->return_size = sz;
+    return 1;
+}
+
+#define PARAM_INT(name, type) \
+    int OSSL_PARAM_get_##name(const OSSL_PARAM *p, const char *key, type *val) \
+    { \
+        return OSSL_PARAM_get_int_common(p, key, val, sizeof(*val)); \
+    } \
+    int OSSL_PARAM_set_##name(const OSSL_PARAM *p, const char *key, \
+                              const type *val) \
+    { \
+        return OSSL_PARAM_set_int_common(p, key, val, sizeof(*val)); \
+    }
+
+PARAM_INT(char, char)
+PARAM_INT(short, short int)
+PARAM_INT(int, int)
+PARAM_INT(long, long int)
+PARAM_INT(long_long, long long int)
+PARAM_INT(int8, int8_t)
+PARAM_INT(int16, int16_t)
+PARAM_INT(int32, int32_t)
+PARAM_INT(int64, int64_t)
+
+PARAM_INT(uchar, unsigned char)
+PARAM_INT(ushort, unsigned short int)
+PARAM_INT(uint, unsigned int)
+PARAM_INT(ulong, unsigned long int)
+PARAM_INT(ulong_long, unsigned long long int)
+PARAM_INT(uint8, uint8_t)
+PARAM_INT(uint16, uint16_t)
+PARAM_INT(uint32, uint32_t)
+PARAM_INT(uint64, uint64_t)
+PARAM_INT(size_t, size_t)
+
+int OSSL_PARAM_get_BN(const OSSL_PARAM *p, const char *key, BIGNUM **val)
+{
+    const union {
+        long one;
+        char little;
+    } is_endian = { 1 };
+    BIGNUM *b;
+
+    if ((p = OSSL_PARAM_locate(p, key)) == NULL)
+        return 0;
+
+    if (p->data_type == OSSL_PARAM_UNSIGNED_INTEGER) {
+        if (is_endian.little)
+            b = BN_lebin2bn(p->buffer, p->buffer_size, *val);
+        else
+            b = BN_bin2bn(p->buffer, p->buffer_size, *val);
+        if (b != NULL) {
+            *val = b;
+            if (p->return_size != NULL)
+                *p->return_size = BN_num_bytes(b);
+            return 1;
+        }
+    }
+    return 0;
+}
+
+int OSSL_PARAM_set_BN(const OSSL_PARAM *p, const char *key, const BIGNUM *val)
+{
+    const union {
+        long one;
+        char little;
+    } is_endian = { 1 };
+    int r;
+    const size_t bytes = (size_t)BN_num_bytes(val);
+
+    if ((p = OSSL_PARAM_locate(p, key)) == NULL)
+        return 0;
+
+    if (p->buffer_size < bytes)
+        return 0;
+
+    if (p->data_type == OSSL_PARAM_UNSIGNED_INTEGER) {
+        if (is_endian.little)
```

Comment:
After all, the whole point of this stuff was to just use the native encoding, no?  BN's are strange and things become simpler if you just use canonical, IMO.

Reply:
What do you mean with "canonical"?

Reply:
Canonical means pick one format and use it, such as always using BN_lebin2bin. As it stands right now, you have to document the format of both BN2bin functions, right?

Reply:
Or simpler, just encode it as a byte array, LSB first, and document that.

Reply:
So basically, you're suggesting that we have two types of integers, one that fits in a long and is native byte order and one that fits in a BIGNUM and always LSB first no matter what?

How does that make anything easier?  'cause then you must suddenly check the size and choose the type depending on size...

Reply:
Either a parameter you're passing is a BIGNUM or it's a scalar. I don't understand why you think it's a problem, the code is going to know that RSA keys take a couple of BIGNUM's, and the output param of an encryption routine will enclude the blocksize or such.

If you really want all those subtypes of integers, shrug, do it but use int64/uint64 as the only thing that goes into a PARAM value.  I would not expose subtypes of integers because I think it imposes too much of a burden on programmers.  I mean really, the API is complex enough and you're adding three-dozen functions to avoid a couple of casts?  That is, seriously, the WRONG trade-off to make here.  It's way to complicated for neglible efficiency.

Reply:
> If you really want all those subtypes of integers

You mistake me for someone else.  I rather care about the `OSSL_PARAM` structure, and the simplicity there is that *all* integers regardless of size are organized the same way.  What you are suggesting leads to having two kinds of integers at that level.

Reply:
The idea here isn't to be clever or to optimise early, it is to allow the type of a parameter to change in the future without breaking existing applications.  No matter how hard we try, types will be incorrect.  Being forced to live with a mistake forever isn't fun.

An example: how many places are there where an _unsigned_ or _size_t_ would make better sense but the API has an _int_?

Another: what would we do if an attack were found against small _e_ values in RSA and a **large** _e_ was suddenly desirable?

A third: what do we do when CPUs introduce native 128 bit integers?


I agree that too many types are supported here at the moment, that doesn't mean only supporting one is ideal.

Reply:
BIGNUM isn't a C integral type; making everything be like a BIGNUM for a mechanism that is a wrapper around C API's seems dumb to me.  I think there should be two types, C numbers and BIGNUM's.  If there is some break in RSA and we have to increase the sizes of e, it will not be binary compatible anyway. When a CPU introduces longlonglong 128bit ints, it won't be source or binary compatible anyway either, no matter how much you try to pretend it will.

Explicitly sized parameters that are internally all treated as 64 bit integers hits the tradeoff among all of flexibility, simplicity, and what we need.

Reply:
This isn't just about us. 

Imagine someone using `uint128_t` in their provider. Should they just do what they usually do with integers, i.e. something like `*(uint128_t *)param->buffer = my128bitint`, or do they have to ensure (possibly convert) that those integers are always whatever we happen to choose (MSB or LSB first), just because we arbitrarily set the limit of acceptable native integers to 64 bits?

Or even better, say the application that wants to use `uint128_t` and just place the address to such variables in the OSSL_PARAM arrays they set up?

I'm starting to wonder if it would help if there were a couple of functions `BN_native2bn` and `BN_bn2native`. I believe that would clear up this discussion.

Reply:
I just submitted #8346, that should make it easier for us.

I just had a look at GMP, and it already has the functionality required, [including support of native endianness](https://gmplib.org/manual/Integer-Import-and-Export.html#Integer-Import-and-Export)

Reply:
"This isn't just about us."  Yes, it is.  If OpenSSL doesn't support 128bit scalar integers, then the provider has to use BIGNUM.  You can't cross into (and out of) OpenSSL without using datatypes that OpenSSL itself understands.

Reply:
So with an agnostic OSSL_PARAM_INTEGER with a well defined form, one end will be able to treat it as a uint128_t while the other end will have to resort to BIGNUM for the same value.

I'm sorry, I don't see the problem there.

Reply:
Seriously though. I don't quite understand the dispute. What's wrong with one unified integer format, regardless of size?

Reply:
You need/want some efficiency, otherwise convert everything to BIGNUM and use that. To not pass an "int" as an "int" is dumb.  You want the param list to mimic what the real API would be; anything else is needless confusion to developers.

This is all just doing procedure calls, through an intermediary, *on the same host* and *in the same process.*

Reply:
??? what stops `param->buffer` from pointing at an `int`

Reply:
... let me just say that `param->buffer` can point at an `int`.
... and also that discussing the pros and cons of raw `OSSL_PARAM` here feels increasingly misplaced.

Reply:
Where do you want the public community to discuss this, if not in the PR?

Reply:
*The* PR?

This PR is about *wrappers* around `OSSL_PARAM`. `OSSL_PARAM` itself comes from #8286 (and even further back, from the design document)

Reply:
This is the first PR that presents the real developer-facing API.  The design document was not public during almost all of its writing.  If you don't like comments here, again I ask where you want them?

Reply:
#8286

Reply:
... If it's about OSSL_PARAM itself...